File: DSA/saumya/DAY8_Q2.py
class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, x):
        self.queue.append(x)

    def dequeue(self):
        if not self.is_empty():
            return self.queue.pop(0)
        return None

    def is_empty(self):
        return len(self.queue) == 0

    def size(self):
        return len(self.queue)

class Stack:
    def __init__(self):
        self.q1 = Queue()
        self.q2 = Queue()

    def push(self, x):
        self.q2.enqueue(x)
        while not self.q1.is_empty():
            self.q2.enqueue(self.q1.dequeue())
        self.q1, self.q2 = self.q2, self.q1

    def pop(self):
        if not self.q1.is_empty():
            return self.q1.dequeue()
        return None

    def top(self):
        if not self.q1.is_empty():
            return self.q1.queue[0]
        return None

#TEST CASE
stack = Stack()
stack.push(1)
stack.push(2)
print(stack.top())  # Output: 2
print(stack.pop())  # Output: 2
print(stack.top())  # Output: 1

File: DSA/Yash_Gupta/Day_8_Q2.java
//Yash gupta 500125397
public class Day_8_Q2 {
    public static void main(String args[])
    {
        //test case-1
        Queue q=new Queue(5);
        Stack s=new Stack();
        s.push(q, 1);
        s.push(q, 2);
        System.out.println(s.pop(q)); 
    }
}
class Stack{
    public void push(Queue q,int num)
    {
        q.enqueue(num);
    }
    public int pop(Queue q)
    {
        Queue temp=new Queue(q.rear+1);
        for(int i=0;i<q.rear;i++)
        {
            temp.enqueue(q.arr[0]);
            q.dequeue();
        }
        int cpy=q.arr[0];
        q.dequeue();
        for(int i=0;i<=temp.rear;i++)
        {
            q.enqueue(temp.arr[0]);
            temp.dequeue();
        }
        return cpy;
    }
    public int top(Queue q)
    {
        return q.arr[q.rear];
    }
}
class Queue{
    int arr[];
    int front;
    int rear;
    Queue(int n){
        this.front=-1;
        this.rear=-1;
        arr=new int[n];
    }
    public void enqueue(int num)
    {
        if(isFull()){return;}
        this.rear++;
        arr[rear]=num;
    }
    public void dequeue()
    {
        if(isEmpty()){return;}
        for(int i=0;i<=rear;i++)
        {
            arr[i]=arr[i+1];
        }
        this.rear--;
    }
    public boolean isEmpty()
    {
        return (rear==-1);
    }
    public boolean isFull()
    {
        return ((rear+1)==arr.length);
    }
    public void printQueue()
    {
        for(int i=0;i<=rear;i++)
        {
            System.out.println(arr[i]);
        }
    }
}
